File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/pigeon_implement/DeviceDelegateImpl.swift
//
//  DeviceDelegateImpl.swift
//  protocol_channel
//
//  Created by hc on 2023/7/24.
//

import Foundation

class DeviceDelegateImpl: DeviceDelegate {
    static let shared = DeviceDelegateImpl()
    private init() {}
    
    private var _deviceInfo: DeviceInfo?
    var deviceInfo: DeviceInfo? {
        return _deviceInfo
    }
    
    var callbackDeviceInfo: ((IDODeviceInterface) -> Void)?
    
    /// 设备信息变更通知
    func listenDeviceChanged(deviceInfo: DeviceInfo) throws {
        _deviceInfo = deviceInfo
    }
    
    /// 绑定时获取的设备信息通知(绑定专用)
    func listenDeviceOnBind(deviceInfo: DeviceInfo) throws {
        callbackDeviceInfo?(DeviceImpl(deviceInfo: deviceInfo))
        callbackDeviceInfo = nil
    }
}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/pigeon_implement/DeviceDelegateImpl.kt
package com.idosmart.pigeon_implement

import com.idosmart.pigeongen.api_device.DeviceDelegate
import com.idosmart.pigeongen.api_device.DeviceInfo
import com.idosmart.protocol_sdk.DeviceInterface

internal class DeviceDelegateImpl: DeviceDelegate {

    companion object {
        @Volatile
        private var instance: DeviceDelegateImpl? = null
        fun instance(): DeviceDelegateImpl {
            return instance ?: synchronized(this) {
                instance ?: DeviceDelegateImpl().also { instance = it }
            }
        }
    }

    private var _deviceInfo: DeviceInfo? = null

    var callbackDeviceInfo: ((DeviceInterface) -> Unit)? = null

    val deviceInfo: DeviceInfo? get() {
        return _deviceInfo
    }

    override fun listenDeviceChanged(deviceInfo: DeviceInfo) {
        _deviceInfo = deviceInfo
    }

    override fun listenDeviceOnBind(deviceInfo: DeviceInfo) {
        val device = IDODevice(deviceInfo)
        callbackDeviceInfo?.invoke(device)
        callbackDeviceInfo = null
    }

}